### ISSUE TITLE
External place for keeping secrets

### ISSUE THREAD

**Issue by iszulcdeepsense** [TheRacetrack/racetrack#164]
Should we consider keeping secrets outside of Kubernetes (Vault, etc)? So that once the kubernetes cluster is upgraded and wiped out, the jobs can be reconciled and revived automatically.

The only reason why jobs are now LOST and can't be recreated is due to secrets are gone.

Let's consider using [Vault by HashiCorp](https://www.vaultproject.io/) or other tool for secrets management.

1. Primary goal is to make all the jobs reproducible so they can be reconciled by Racetrack without constantly asking users to redeploy.
2. We can also consider the new place for secrets to be utilized directly by users (job developers). So they can push their git credentials and runtime environment variables to the Vault in first place and to stop keeping their secrets in the local `.env` files.

**Comment by JosefAssadERST** [comment]
Right, so some unstructured and unordered thoughts.

* The core idea is to outsource secrets management to a system off Kubernetes, Vault being the obvious candidate. Whether the secret owners populate them directly in Vault or keep putting them in env files is a secondary matter.
* Does this create a hard dependency on Vault? I'm concerned bout that since Vault is [open core](https://en.wikipedia.org/wiki/Open-core_model). Can such a dependency be softened by a "use Vault if one is available, otherwise don't use Vault"? If yes, is it a lot of code to maintain to soften in this way?
* Is Vault the only real game in town? Open Core isn't the only point of concern; a competitor might be fully FOSS but 99.99% of commits come from one commercial entity, raising the risk profile.
* If we implement this and allow users to enter secrets in Vault, will they think it's nice on day 0 and get annoyed by the extra step on day 100? The truth is, this would mean job developers have two places to type the right thing not just one: the git repo, and in Vault.

@jsandroos @LookACastle @mariushart @Koshmaar thoughts?

**Comment by LookACastle** [comment]
I know of no option that's more "free as in freedom" than hashicorp vault's open source offering, they're the only real game in town I know of that's any degree of free.

If it's doable to support a vault without breaking the option of not having one, then I'd argue we don't let perfect be the enemy of good here. K8s has a decently quick release cadence, if people upgrade whenever a new version is released, they'd end up having to deal with lost jobs fairly often. 

If both having and not having a vault is supported, then the concern of annoyed developers also kind of disappears - they can just not use the feature if they don't like it.

**Comment by Valleo4** [comment]
data engineering has long had an issue for implementing a vault for secrets [here](https://rep.erst.dk/git/ml-labs/pipelines/-/issues/69), since we handle them in .env files, which is not ideal.
so if you consider doing a Vault we would be interested in using it as well :)

**Comment by Koshmaar** [comment]
It seems for me the option B would fit here more. But a tangential issue, is that perhaps we could avoid storing user specific secrets (like their gitlab tokens), if we could use service accounts. I heard one secret data scientists store in Jobs, is their gitlab tokens, to access private repositories. Probably you have already considered using gitlab Project Access Tokens. What was the reason against?

Also regarding the option B, probably we could set in Vault policy which says that user logged using token ABC, can access only secret coupled with it, ie.
1. Job Foobar in .env stores `vault_token: ABC, vault_path: foobar/secret`
2. Vault admin sets a key `foobar/secret` with content `123` (the secret which is needed by Job to do its.. job, ie. db password), and sets policy to be accessible only with token ABC.
3. Job Foobar when deployed, in init func uses http with token ABC, to fetch from `vault_path` the secret `123`.

That way leaking single password to Vault, won't leak all others. Which is typical problem in this area:

![image](https://user-images.githubusercontent.com/5213082/219477176-ab5809b6-db45-439a-86d7-4491bb8e4dba.png)

There are some approaches they recommend: https://developer.hashicorp.com/vault/tutorials/app-integration/secure-introduction?in=vault%2Fapp-integration

--------------------

Btw, I was lucky when doing googling, and found this unpolished gem (or poo, you decide): http://assimilationsystems.com/2020/09/01/sharing-secrets-with-containers-using-custodia/
It shows how adding of overseeing component on k8s server helps to avoid storing any secrets on the client, by means of identifying the process that tries to read the token, through linux socket proc info (you could think of it as "biometric" authorization). The software is here: https://github.com/latchset/custodia  On gpl 3, but a bit forgotten and complex, so not ideal for integration. 

But there's hope, as the guy wrote a second article with an even simpler method without messing with k8s internals:  http://assimilationsystems.com/2020/09/01/the-authproxy-method-of-sharing-secrets-safely-with-containers/   It requires to have Vault, but all secrets are stored there, none need to be stored on disk (of Job). There's little piece of code (authproxy) which listens on socket, identifies caller (ie. using process name), and for authorized, asks the Vault using its embedded secret, for required key and sends it to app/Job directly (could be checking first that ie. docker container with certain hash listens on certain port). Please read the article for more info :) Seems rather secure, and could fit Racetrack model (Jobs could have plugin which provides such authproxy container next to Job), without making Vault hard dependency. There's no reference implementation and it's not trivial thing, but logic is straightforward, and should be implementable by anyone... at least that's my initial impression.

Sorry for wall of text, and hope this will be of any use.

**Comment by LookACastle** [comment]
Authproxy is a really interesting concept, great find! I'd be amiss if I didn't note that there's a known race condition that the article mentions and gives a few links to discussion on. My conclusion after reading is that it's a purely theoretical exploit and not worth worrying that much about, but we need to be aware that it exists.

**Comment by iszulcdeepsense** [comment]
In Racetrack, infrastructure plugins are responsible for taking care of secrets right now:
https://github.com/TheRacetrack/racetrack/blob/master/lifecycle/lifecycle/deployer/deploy.py#LL49C3-L53C3

As discussed, we can make Vault functionality baked in the Racetrack core as a configurable option.
This will make it less intricate in case of Kubernetes plugin + Vault enabled.

**Comment by JosefAssadERST** [comment]
Here's a concrete proposal. Tell me how much of it you object to.

1. Bake Hashicorp Vault in to RT core.
2. Remove secrets management entirely from infrastructure target plugins (hereafter ITP)

# Advantages:

1. RT admins don't have to do a lot of thinking about where a secret for a given job might be, or whether installing a new ITP will implicitly override anything or add new options somewhere (e.g. if that plugin has its own secrets management backend)
2. Secrets can be populated before the ITP is even installed (weird scenario, but people are often weird)
3. Fewer if/elses in core and potentially in plugins
4. It's probably the simplest solution that works.
5. Reduces the footprint of ITPs since it means they should explicitly not do secrets management any longer.

# Disadvantages

1. It's open core not open source. We've documented our concern
2. Bloats RT core (a bit?)
3. Not the most "acrobatic" solution: it would be more flexible if an infrastructure type could offer an alternative secrets management backend, but: the truth is, we don't _need_ different secrets management backends. We just need one that works for any ITP.

**Comment by iszulcdeepsense** [comment]
@JosefAssadERST I generally agree. Just wondering about one thing: what about keeping secrets in k8s? Do we want to completely drop it in favour of Vault?

**Comment by JosefAssadERST** [comment]
That's the idea. We don't know if in ten years 99% of our jobs are served in VMWare or on docker daemons. Vault is a nutral place that my proposal intends to work regardless of infrastructure type.

**Comment by iszulcdeepsense** [comment]
Good. It also feels more natural to take secrets away from infrastructure plugins, especially when some of them (ie. docker infrastructure) don't implement secrets at all.

**Comment by JosefAssadERST** [comment]
I have no fundamental disagreements. Still, there's some really big advantages to splitting off secrets management. Can you suggest an alternative?

**Comment by LookACastle** [comment]
The alternative is supporting vault, without having it as explicitly the only option. Keep the old secrets management as well - or rip it out, but provide vault as some sort of plugin so we could at least _in theory_ not have a vault if it starts becoming a problematic piece of software for us.

**Comment by JosefAssadERST** [comment]
I am also in favor of making Vault optional, but how would that work assuming we rip secrets management our of infrastructure target plugins?

Would there simply not be any support for secret management?

**Comment by LookACastle** [comment]
If it's deemed too difficult to make secrets management plugins, then I'd rather keep secrets management in ITPs. Whether that be a load of ITPs like `plugin-kubernetes-infrastructure-vault`, `plugin-kubernetes-infrastructure-k8s-secrets`, and so on (no matter how inelegant I find that solution), or a big bloated plugin that supports all the secrets. I'd rather have either than I'd have us explicitly tied to COSS.

I don't think (knock on wood) that secrets management plugins would be too difficult to make though. Getting the secrets securely is the hard part - and we'd have to do that anyhow - after that they should just be exposing data within the walled garden, no? It's not like a secret pulled from k8s is gonna be different from one pulled from vault - they have the same value.

If we do make secrets management plugins, the default would probably be no support for secrets without a plugin though, yes.

**Comment by LookACastle** [comment]
Actually, we could also add all the secrets stuff to racetrack core probably, but I still think that having it as plugins makes the most sense.

**Comment by iszulcdeepsense** [comment]
I don't have a strong opinion on your proposals, both sound reasonable. Just quickly checked that having secrets plugins  wouldn't be too much of a burden on code complexity as we anticipated, there's only 2 places: [saving](https://github.com/TheRacetrack/racetrack/blob/master/lifecycle/lifecycle/deployer/deploy.py#L51) and [reading](https://github.com/TheRacetrack/racetrack/blob/master/lifecycle/lifecycle/deployer/redeploy.py#L134). 
It will just proliferate plugins even more. I wonder if it's good or bad. It's good in terms of various possibilities, it's bad in terms of installing them, upgrading and maintaining the code.

**Comment by JosefAssadERST** [comment]
How about:

* Rip secrets management out of infr.t. plugins
* Bake Hashicorp Vault into core
* Make it optional to use; if there's no Vault, RT works just as well but does not handle secrets, they get hardcoded into the jobs by job developers
* Later if we find a good reason, we can make it a plugin
* Later if a real FOSS alternative emerges, we can migrate to it

?

**Comment by LookACastle** [comment]
Good enough for me, not perfect - I'd prefer to just have it as a plugin - but perfect is the enemy of good. If it's easier to bake support into the core rather than start off with a plugin, I'd say that's the correct way to do it until something comes along to change our mind. YAGNI and all that.

**Comment by iszulcdeepsense** [comment]
Agree. Let's not fall into YAGNI, though I believe Vault can be "softly hardcoded" by making it as generic as possible (separate module, etc), having in mind a possible change in future.

**Comment by JSACol** [comment]
Le Dot.

Did we ever move forward with a branch/PR for this?

**Comment by LookACastle** [comment]
Hashicorp functionally shot Vault dead, so we had nothing to develop against for a while. I think maybe OpenBao exists now in a complete enough state to be useful? There's been some murmurs at least, it's worth it to look into whether we can continue working on the issue.

**Comment by iszulcdeepsense** [comment]
Right, probably OpenBao would be the best fit at this moment. It got stuck due to uncertain situation with Hashicorp Vault.

**Comment by JSACol** [comment]
> Hashicorp functionally shot Vault dead

For sure no one is going to contribute to a software project they are not allowed to use. It is basically the OpenOffice/LibreOffice story repeated again. Libre office is alive and thriving today and OpenOffice is dead. So there is a bit of positive precedence in cases like this.

Anyhow, You're thinking RT would be considered a competitor under the Business License Hashicorp is now using? 
https://www.hashicorp.com/en/license-faq

Otherwise I agree OpenBao would be a good alternative (or continuation)

Cheers
   Joakim
